tracer/src/Datadog.Trace/RemoteConfigurationManagement/RemoteConfigurationManager.cs

```diff
@@ -0,0 +1,272 @@
+                }
+
+                try
+                {
+                    await Task.Delay(_pollInterval, _cancellationSource.Token).ConfigureAwait(false);
+                }
+                catch (TaskCanceledException)
+                {
+                    // We are shutting down, so don't do anything about it
+                }
+            }
+        }
+
+        public void RegisterProduct(Product product)
+        {
+            _products.TryAdd(product.Name, product);
+        }
+
+        public void UnregisterProduct(string productName)
+        {
+            _products.TryRemove(productName, out _);
+        }
+
+        private async Task Poll()
+        {
+            await SemaphoreSlim.WaitAsync().ConfigureAwait(false);
+
+            try
+            {
+                var request = BuildRequest();
+                var response = await _remoteConfigurationApi.GetConfigs(request).ConfigureAwait(false);
+
+                if (response?.Targets?.Signed != null)
+                {
+                    ProcessResponse(response);
+                    _targetsVersion = response.Targets.Signed.Version;
+                }
+            }
+            catch (Exception e)
+            {
+                _lastPollError = e.Message;
+            }
+            finally
+            {
+                SemaphoreSlim.Release();
+            }
+        }
+
+        private GetRcmRequest BuildRequest()
+        {
+            var appliedConfigurations = _products.Values.SelectMany(pair => pair.AppliedConfigurations.Values);
+
+            var cachedTargetFiles = new List<RcmCachedTargetFile>();
+            var configStates = new List<RcmConfigState>();
+
+            foreach (var cache in appliedConfigurations)
+            {
+                cachedTargetFiles.Add(new RcmCachedTargetFile(cache.Path.Path, cache.Length, cache.Hashes.Select(kp => new RcmCachedTargetFileHash(kp.Key, kp.Value)).ToList()));
+                configStates.Add(new RcmConfigState(cache.Path.Id, cache.Version, cache.Path.Product));
+            }
+
+            var rcmState = new RcmClientState(_rootVersion, _targetsVersion, configStates, _lastPollError != null, _lastPollError);
+            var rcmClient = new RcmClient(_id, _products.Keys, _rcmTracer, rcmState);
+            var rcmRequest = new GetRcmRequest(rcmClient, cachedTargetFiles);
+
+            return rcmRequest;
+        }
+
+        private void ProcessResponse(GetRcmResponse response)
+        {
+            var actualConfigPath =
+                response
+                   .ClientConfigs
+                   .Select(RemoteConfigurationPath.FromPath)
+                   .ToDictionary(path => path.Path);
+
+            var changedConfigurationsByProduct = GetChangedConfigurations()
+               .GroupBy(config => config.Path.Product);
+
+            foreach (var productGroup in changedConfigurationsByProduct)
+            {
+                var product = _products[productGroup.Key];
```

Comment:
Is there a risk here that `productGroup.Key` might not exist in `_products`, especially considering we can arbitrarily add and remove products?

Reply:
No, there is no such risk, since this checked at `GetChangedConfigurations()`

https://github.com/DataDog/dd-trace-dotnet/blob/66f71cfe08ba827242cd91ab849b3f724c21de9c/tracer/src/Datadog.Trace/RemoteConfigurationManagement/RemoteConfigurationManager.cs#L220

Reply:
Changed to copied dictionary here 8236f872cd3549d0f69eb0caf3704ea44d4222ca